### DISCUSSION TITLE
AWS SAM JSON Schema

### DISCUSSION THREAD

**Discussion by hoffa** [aws/serverless-application-model#2645] [Announcements]
## Introduction

We now provide a [JSON Schema](https://json-schema.org/) for [AWS SAM templates](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification-template-anatomy.html) (see https://github.com/aws/serverless-application-model/issues/1133 for the original issue). It enhances the template authoring experience by enabling autocompletion, validation, and viewing documentation directly from your IDE. It supports all SAM and [AWS CloudFormation resources](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-template-resource-type-ref.html).

Here's what it looks like in Visual Studio Code:

<p align="center">
<img width="756" alt="screen_shot_2022-11-24_at_13 59 57" src="https://user-images.githubusercontent.com/1280602/203868436-9ce5c971-d36b-4f67-9dfd-9d43ef94c5cb.png">
</p>

## JSON Schema files

> **Warning**
> The schema is **over 15 MB**.
> If you're integrating into other software, we highly recommend downloading the compressed schema instead. For example:
>
> ```
> curl -H 'Accept-Encoding: gzip' https://raw.githubusercontent.com/aws/serverless-application-model/main/samtranslator/schema/schema.json | gunzip
> ```

- [Latest stable AWS SAM JSON Schema](https://github.com/aws/serverless-application-model/blob/main/samtranslator/schema/schema.json)
- [Latest development AWS SAM JSON Schema](https://github.com/aws/serverless-application-model/blob/develop/samtranslator/schema/schema.json)

### Using AWS Toolkit for Visual Studio Code

Install [AWS Toolkit for Visual Studio Code](https://aws.amazon.com/visualstudiocode/) and open any SAM template. It uses the SAM JSON Schema under the hood.

> **Note**
> There's a known performance issue with documentation tooltips in AWS Toolkit for Visual Studio Code. See https://github.com/aws/aws-toolkit-vscode/issues/3259. If it's an issue, you can use the methods below as workaround.

### Using the Visual Studio Code YAML extension

1. Install the [YAML extension](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml).
2. Paste the following at the top of your SAM template:
 
    ```yaml
    # yaml-language-server: $schema=https://raw.githubusercontent.com/aws/serverless-application-model/main/samtranslator/schema/schema.json
    ```

### Using Visual Studio Code JSON Schema support

If your template is in JSON format, add the following at the root of the template:

```json
{
  "$schema": "https://raw.githubusercontent.com/aws/serverless-application-model/main/samtranslator/schema/schema.json"
}
```

## Technical details

See [here](https://github.com/aws/serverless-application-model/blob/develop/docs/schema.md) for more details.

## Limitations

1. **Limited support for [intrinsic functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html).** Schema validation errors are expected when intrinsic functions are used. See https://github.com/aws/aws-toolkit-vscode/issues/2596.
2. **Only supports templates understood by the SAM transform.** A template won't necessarily validate against the schema if it includes other [AWS CloudFormation macros](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html) that modify the template. For example, the template in https://github.com/aws/serverless-application-model/issues/1859#issuecomment-1346475672 doesn't pass schema validation since it uses `AWS::LanguageExtensions` to modify properties of a SAM resource.
3. **Limited support for SAM pass-through values.** Pass-through properties (e.g. [`AccessLogSettings`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-httpapi.html#sam-httpapi-accesslogsettings)) are passed directly to underlying CloudFormation resources. This is being addressed in https://github.com/aws/serverless-application-model/pull/3055.

**Comment by CGarces** [comment]
I get some error on my SAM templates and the tooltip not provide information.
I'ts something wrong in my environment?

![image](https://user-images.githubusercontent.com/6617155/203957363-733b1145-4175-4846-ae4b-f226ae795c18.png)

![image](https://user-images.githubusercontent.com/6617155/203957327-db105669-1d43-406e-ac09-71e159b0b2d4.png)

**Reply by hoffa** [reply]
Interesting. The error message isn't very helpful, but I'm guessing the `sam.schema.json` refers to https://github.com/awslabs/goformation/blob/master/schema/sam.schema.json (another SAM schema, not from the SAM team) which is used among other places in the AWS Toolkit (see [here](https://github.com/aws/aws-toolkit-vscode/blob/3090e4fee6854b6de3d8a716bfefa10d5ad88b2c/src/shared/schemas.ts#L167)) and the JSON Schema Store (see [here](https://github.com/SchemaStore/schemastore/blob/0b787a13ae44d17173b9368c50727bd33ed7c17e/src/api/json/catalog.json#L995)).

**Reply by hoffa** [reply]
Was able to reproduce after installing [AWS Toolkit](https://marketplace.visualstudio.com/items?itemName=AmazonWebServices.aws-toolkit-vscode). Looks like the schema features it ships as well are somehow interfering. We'll look into this, thanks for the heads-up!

Edit: see https://github.com/aws/aws-toolkit-vscode/issues/3012

**Reply by jpinkney-aws** [reply]
We've fixed this in the VSCode-YAML extension so when users do:
```
# yaml-language-server: $schema=https://raw.githubusercontent.com/aws/serverless-application-model/main/samtranslator/schema/schema.json
```
at the top of their file it should now pick up the newly create sam schema instead! You just need to ensure that you have VSCode-YAML 1.11 and above

**Comment by justinmk3** [comment]
it should also be possible to manually select the json schema from the vscode status bar, but it seems like that is not working. tracked in https://github.com/aws/aws-toolkit-vscode/issues/3012

![image](https://user-images.githubusercontent.com/55561878/204043632-46d1edf9-7a8f-406f-a3ef-5cc8bfdc9a58.png)

**Comment by SimonCMoore** [comment]
We are looking into a resolution but if you want a temporary workaround should you encounter the above issue, locate AWS Toolkit in your installed packages and you can temporarily disable, reload vscode and you can then use the schema as described.

**Comment by ryanotella** [comment]
_IntelliJ_  
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/276385/205039164-ab9ec859-e608-4972-aaf7-b3360b53c43f.png">

**Comment by HanaaGhasoub** [comment]
How can I reference to external file has the definition of the resources in SAM template with json format
<img width="293" alt="image" src="https://user-images.githubusercontent.com/30862539/225252943-95968fda-d0ed-4f2f-804d-d1523da31a2a.png">

**Reply by hoffa** [reply]
The JSON schema doesn't help with referencing external files, it only helps with the development experience of reading and writing valid templates by providing things like autocompletion and documentation.

To answer your question however, you should be able to include JSON files as such:

```yaml
Fn::Transform:
  Name: AWS::Include
  Parameters:
    Location: <path-to-json>
```

For example:

```yaml
Transform: AWS::Serverless-2016-10-31
Resources:
  Fn::Transform:
    Name: AWS::Include
    Parameters:
      Location: resources.json
```

Where `resources.json` is:

```json
{
  "MyTable": {
    "Type": "AWS::Serverless::SimpleTable"
  }
}
```

You'll need to use the SAM CLI to deploy, so that it can [package](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-package.html) the assets for you.

If the issue is with the error in the IDE (but no issue with deployments), if you can specify what the error is and the parent keys of `Resources` (to understand where in the template it shows up), I can check in the schema and update if it's a type that should be valid (within the limitations mentioned in https://github.com/aws/serverless-application-model/discussions/2645#discussion-4591631).

**Reply by HanaaGhasoub** [reply]
@hoffa Thank you so much, that's pretty explanation.  I will follow your instructions and update you with result.
